simularium/simularium-website
# Feature/color session

Time estimate or Size
=======
_large_

I split some of the changes that were pure renaming/organization into a leading PR to ease this one, but there is still quite a bit of redux boiler plate, as well as tests that inflate the number of lines here.

It's not a small PR either way, but not quite as bad as it seems.

Problem
=======
Users want to store their color selections for later use, this solves one of two approaches to the issue, which is to store those selections in local browser storage.

Closes #511 and advances #510, design review complete.

Solution
========
#573 makes gets some organizing and renaming out of the way, pointing at that branch, the feature is here.

This branch:
 - `currentColorSettings` enum member `Default` or `UserSelected` indicates which settings should be applied
 - upon parsing a trajectory store the initial result of `getUIDisplayData` as the default settings
  - default UI data lives in `trajectory` branch and settings and user selected UI data live in `selection`
- `selection` branch logics manage putting user selections into and out of local storage in the browser
- file names and data structures are compared before applying any stored settings, to prevent applying settings in cases where file names match but data structures do not: util `compareAgentTrees`
- several asynchronous processes have bearing on our ability to compare these data structures, so some logic for handling that lives in more than one place
- in response to user color selections made via colorPicker: update both user selections redux and local storage for this trajectory
- whenever user color selections are applied, expose a "restore defaults" button on the left side panel that previews the default when hovered over, and restores default and clears user selections from redux and local storage when clicked

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Make selections via color picker and see that they are applied correctly
2. Load a new trajectory, then return to the previous, selections should still be applied
3. Reload the tab, selections should still be applied.
4. Open the trajectory in a new tab and the same settings should still be applied
5. When defaults are applied "Restore default colors" button should be disabled
6. When active, hovering over "Restore default colors" should temporarily apply default settings, and removing hover should go back to user selections
7. Clicking on "Restore default colors" button should permanently apply defaults, clear out previous selections, and disable button
8. Extra credit: download a trajectory, "Example 1", and change its name to match a networked trajectory "Example 2". Store settings for original "Example 2", then load renamed version that is called "Example 2" but is actually "Example 1" under the hood. Settings should not be applied!
